pkg/virt-controller/watch/vm.go

```diff
@@ -576,6 +577,73 @@ func (c *VMController) updatePVCMemoryDumpAnnotation(vm *virtv1.VirtualMachine)
 	return nil
 }
 
+func (c *VMController) saveCurrentVMICPUCsPatch(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
+
+	vmiCopy := vmi.DeepCopy()
+	vmiCopy.Status.CurrentCPUTopology = vmi.Spec.Domain.CPU
+
+	oldJson, err := json.Marshal(vmiCopy.Status.CurrentCPUTopology)
+	if err != nil {
+		return err
+	}
+
+	patchOps := fmt.Sprintf(`{ "op": "add", "path": "/status/currentCPUTopology", "value": %s}`, string(oldJson))
+	patchBytes := controller.GeneratePatchBytes([]string{patchOps})
+
+	if len(patchBytes) == 0 {
+		return nil
+	}
+
+	_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, patchBytes, &v1.PatchOptions{})
+	if err != nil {
+		return err
+	}
+
+	var cpu virtv1.CPU
+	cpu.Sockets = vm.Spec.Template.Spec.Domain.CPU.Sockets
+	cpu.Cores = vm.Spec.Template.Spec.Domain.CPU.Cores
+	cpu.Threads = vm.Spec.Template.Spec.Domain.CPU.Threads
+	cpu.MaxSockets = vmi.Spec.Domain.CPU.MaxSockets
+
+	newJson, err := json.Marshal(cpu)
+	if err != nil {
+		return err
+	}
+
+	test := fmt.Sprintf(`{ "op": "test", "path": "/spec/domain/cpu", "value": %s}`, string(oldJson))
+	update := fmt.Sprintf(`{ "op": "replace", "path": "/spec/domain/cpu", "value": %s}`, string(newJson))
+	patch := fmt.Sprintf("[%s, %s]", test, update)
+	_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, []byte(patch), &v1.PatchOptions{})
+	return err
+}
+
+func (c *VMController) handleCPUChangeRequest(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
+
+	if vm.Spec.LiveUpdateFeatures == nil {
+		return nil
+	}
+
+	if vmi == nil || vmi.DeletionTimestamp != nil {
+		return nil
+	}
+
+	if vm.Spec.Template.Spec.Domain.CPU == nil || vmi.Spec.Domain.CPU == nil {
+		return nil
+	}
```

Comment:
Isn't it enough to only check the VM?
I mean, is it even possible that only one of them is nil?